### PR TITLE
Implement the training of FormatAnalyzer

### PR DESCRIPTION
**Pull request by m09**
The train method uses model selection to obtain the best possible set
of rules with a non destructive pruning method. We can then implement
destructive pruning if we need to further reduce the number of rules.

Signed-off-by: Hugo Mougard <hugo@sourced.tech>

### CODE REVIEW

File: lookout/style/format/analyzer.py [hunk 1]
```diff
@@ -1,19 +1,25 @@
...
 from lookout.style.format.model import FormatModel
+from lookout.style.format.rules import TrainableRules
 
 
 class FormatAnalyzer(Analyzer):
     log = logging.getLogger("FormatAnalyzer")
     model_type = FormatModel
     version = "1"
     description = "Source code formatting: whitespace, new lines, quotes, braces."
+    LOWER_BOUND_INSTANCES = 500
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: All the configuration is built in `config` in `train()`.

File: lookout/style/format/analyzer.py [hunk 2]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+        :param data: data sent by the lookout server
+        :return: a modelforge.Model containing the learned rules, per language.
+        """
         cls.log.info("train %s %s %s", ptr.url, ptr.commit, data)
         files = data["files"]
+        language_to_files = defaultdict(list)
         for file in files:
+            if not len(file.uast.children):
+                continue
+            # TODO: replace this temporary hack when langauge fields in File objects are fixed
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: I have added the TODO format: https://github.com/src-d/guide/blob/d76e64ae32a9afbdfcf8fbefdecf683459176d79/engineering/conventions-python.md#code-style

Copied from Google, this allows asking the person who wrote the TODO what he meant without running `git blame`

File: lookout/style/format/analyzer.py [hunk 3]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
-        4. extract_features()
-        5. train the base model - tree or forest
-        6. train the rules
-        7. generate the Modelforge model
-        """
+        model = FormatModel().construct(cls, ptr)
 
-        return FormatModel().construct(cls, ptr)
+        for language, files in language_to_files.items():
+            cls.log.info("processing %s files", language)
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: training on %d files of language %s

File: lookout/style/format/analyzer.py [hunk 4]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+            cls.log.info("extracting features from %d files", len(files))
+            fe = FeatureExtractor(language=language, siblings_window=5, parents_depth=2)
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: `siblings_window` and `parents_depth` go to `config.get()`

File: lookout/style/format/analyzer.py [hunk 5]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+            X, y = fe.extract_features(files)
+            if X.shape[0] < cls.LOWER_BOUND_INSTANCES:
+                cls.log.warning("did not train a model for language %s: not sufficiently many "
+                                "instances (%d)" % (language, X.shape[0]))
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: We seem to miss `continue` here

Comment (vmarkovtsev) [reply lookout/style/format/analyzer.py]: `"skipped %s: too few samples (%d < %d)" % (file.path, X.shape[0], min_samples)`

File: lookout/style/format/analyzer.py [hunk 6]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+        """
+        Train a model given the files available.
+
+        :param url: url of the repository
+        :param commit: hash of the commit that we analyze
+        :param config: configuration dict
+        :param data: data sent by the lookout server
+        :return: a modelforge.Model containing the learned rules, per language.
+        """
         cls.log.info("train %s %s %s", ptr.url, ptr.commit, data)
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: would be nice to print the config here

File: lookout/style/format/analyzer.py [hunk 7]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
-        return FormatModel().construct(cls, ptr)
+        for language, files in language_to_files.items():
+            cls.log.info("processing %s files", language)
+            cls.log.info("extracting features from %d files", len(files))
+            fe = FeatureExtractor(language=language, siblings_window=5, parents_depth=2)
+            X, y = fe.extract_features(files)
+            if X.shape[0] < cls.LOWER_BOUND_INSTANCES:
+                cls.log.warning("did not train a model for language %s: not sufficiently many "
+                                "instances (%d)" % (language, X.shape[0]))
+            cls.log.info("training rules model")
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: log.debug

File: lookout/style/format/analyzer.py [hunk 8]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+            bscv = BayesSearchCV(
+                TrainableRules(prune_branches_algorithms=["reduced-error"], prune_attributes=False,
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: all these parameters go to `config.get()`

File: lookout/style/format/analyzer.py [hunk 9]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+                 "max_depth": Categorical([None, 5, 10]),
+                 "max_features": Categorical([None, "auto"]),
+                 "min_samples_split": Integer(2, 20),
+                 "min_samples_leaf": Integer(1, 20)},
+                n_jobs=-1,
+                n_iter=30)
+            bscv.fit(X, y)
+            cls.log.info("score of the best estimator found: %.3f", bscv.best_score_)
+            cls.log.info("params of the best estimator found: %s", str(bscv.best_params_))
+            cls.log.info("training the model with complete data")
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: log.debug for all the three

File: lookout/style/format/analyzer.py [hunk 10]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+            trainable_rules = TrainableRules(prune_branches_algorithms=["reduced-error"],
+                                             prune_attributes=True, random_state=42,
+                                             uncertain_attributes=True, **bscv.best_params_)
+            trainable_rules.fit(X, y)
+            cls.log.info("saving %s rules model", language)
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: we don't really do anything special - take the `rules` attribute and put it into the model. can be removed.

File: lookout/style/format/analyzer.py [hunk 11]
```diff
@@ -35,24 +41,61 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+            model[language] = trainable_rules.rules
+        return model
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: log.info("trained %s", model)

File: lookout/style/format/rules.py [hunk 12]
```diff
@@ -155,18 +156,19 @@ class TrainableRules(BaseEstimator, ClassifierMixin):
 
     _log = logging.getLogger("TrainableRules")
 
-    def __init__(self, base_model_name: str, prune_branches=True,
-                 prune_branches_algorithm="top-down-greedy",
+    def __init__(self, base_model_name: str,
+                 prune_branches_algorithms=["reduced-error", "top-down-greedy"],
```
Comment (vmarkovtsev) [inline lookout/style/format/rules.py]: Please make it a tuple (you know, those default lists...)

File: lookout/style/format/rules.py [hunk 13]
```diff
@@ -155,18 +156,19 @@ class TrainableRules(BaseEstimator, ClassifierMixin):
...
+                 random_state=42):
         """
         Initializes a new instance of Rules class.
 
         :param base_model_name: fully qualified type name of the base model to train. \
                                 Must be either "sklearn.tree.DecisionTreeClassifier" or \
                                "sklearn.ensemble.RandomForestClassifier".
-        :param prune_branches: indicates whether to remove useless rules.
-        :param prune_branches_algorithm: chooses the pruning algorithm.
+        :param prune_branches_algorithms: branche pruning algorithms to use.
```
Comment (vmarkovtsev) [inline lookout/style/format/rules.py]: branche -> branch

**Comment by m09** [comment]
about the `log_level` that we should add to TrainableRules, do you mean so that we can `setLevel` to ignore anything below INFO for example? Because that seems possible through regular external logger configuration.

File: lookout/style/format/analyzer.py [hunk 14]
```diff
@@ -35,24 +40,83 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+        """
+        Train a model given the files available.
+
+        :param url: url of the repository
+        :param commit: hash of the commit that we analyze
+        :param config: configuration dict
+        :param data: data sent by the lookout server
+        :return: a modelforge.Model containing the learned rules, per language.
+        """
+        merged_config = {
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: Please name it "final_config" - a merge operation is supposed to be two-way.

File: lookout/style/format/analyzer.py [hunk 15]
```diff
@@ -35,24 +40,83 @@ def analyze(self, ptr_from: ReferencePointer, ptr_to: ReferencePointer,
...
+            "prune_attributes": False,
+            "uncertain_attributes": True,
+            "prune_dataset_ratio": .2,
+            "n_estimators": 10,
+            "random_state": 42,
+            "n_iter": 30
+        }
+        merged_config.update(config)
+        cls.log.info("train %s %s %s with config %s", ptr.url, ptr.commit, data,
+                     str(merged_config))
```
Comment (vmarkovtsev) [inline lookout/style/format/analyzer.py]: No need to `str()` the config. At the same time the order of keys will be random by default, so:

```python
from pprint import pformat
% pformat(final_config, width=4096, compact=True)
```

**Comment by m09** [comment]
I agree that it's nice to have different output levels. WDYT about using an external config to achieve so like https://github.com/m09/sumexp/blob/master/logging.yml?

**Comment by vmarkovtsev** [comment]
We always write to stdout so in our case the role of config is solely setting the log levels. Yep, would be cool to have it.

**Comment by m09** [comment]
I put the logging part in a separate PR (#67)

**Pull request merged**